navigation-material-3/src/main/java/com/google/accompanist/navigation/material3/ModalBottomSheet.kt

```diff
@@ -0,0 +1,131 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:OptIn(ExperimentalMaterial3Api::class)
+
+package com.google.accompanist.navigation.material3
+
+import androidx.compose.material3.BottomSheetDefaults
+import androidx.compose.material3.ExperimentalMaterial3Api
+import androidx.compose.material3.ModalBottomSheet
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateListOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.rememberSaveableStateHolder
+import androidx.compose.runtime.snapshots.SnapshotStateList
+import androidx.compose.ui.platform.LocalInspectionMode
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.navigation.NavBackStackEntry
+import androidx.navigation.compose.LocalOwnersProvider
+
+@OptIn(ExperimentalMaterial3Api::class)
+@ExperimentalMaterial3NavigationApi
+@Composable
+fun ModalBottomSheet(
+    navigator: BottomSheetNavigator,
+    content: @Composable () -> Unit
+) {
+    content()
+
+    val saveableStateHolder = rememberSaveableStateHolder()
+    val sheetBackStack by navigator.backStack.collectAsState()
+    val visibleBackStack = rememberVisibleList(sheetBackStack)
+    visibleBackStack.PopulateVisibleList(sheetBackStack)
+
+    visibleBackStack.forEach { backStackEntry ->
+        val destination = backStackEntry.destination as BottomSheetNavigator.Destination
+        // while in the scope of the composable, we provide the navBackStackEntry as the
+        // ViewModelStoreOwner and LifecycleOwner. We provide this to the whole ModalBottomSheet
+        // because its content has a ColumnScope receiver which we have to honor when composing it.
+        backStackEntry.LocalOwnersProvider(saveableStateHolder) {
+            ModalBottomSheet(
+                onDismissRequest = { navigator.dismiss(backStackEntry) },
+                shape = destination.sheetProperties.shape,
+                containerColor = destination.sheetProperties.containerColor,
+                contentColor = destination.sheetProperties.contentColor,
+                tonalElevation = destination.sheetProperties.tonalElevation,
+                scrimColor = destination.sheetProperties.scrimColor,
+                dragHandle = destination.dragHandle ?: { BottomSheetDefaults.DragHandle() }
+            ) {
+                DisposableEffect(backStackEntry) {
+                    onDispose {
+                        navigator.onTransitionComplete(backStackEntry)
+                    }
+                }
+
+                destination.content(this, backStackEntry)
+            }
+        }
+    }
+}
+
+@Composable
+internal fun MutableList<NavBackStackEntry>.PopulateVisibleList(
```

Comment:
I'm not fully what these functions do together but in the end, this didn't work for me - poping back stack wasn't animating the bottomsheet down. I've tried reimplementing the "retainedBackStackEntry" approach here and I've ended up with this:

```kotlin
@Composable
public fun ModalBottomSheet(
  navigator: BottomSheetNavigator,
) {
  val saveableStateHolder = rememberSaveableStateHolder()

  val sheetBackStackState by remember {
    navigator.backStack.zipWithPreviousCount()
  }.collectAsState(Pair(0, emptyList()))

  val count by remember {
    derivedStateOf {
      maxOf(
        sheetBackStackState.first,
        sheetBackStackState.second.count(),
      )
    }
  }

  repeat(count) { i ->
    val backStackEntry = sheetBackStackState.second.getOrNull(i)
    ModalBottomSheet(
      navigator = navigator,
      saveableStateHolder = saveableStateHolder,
      targetBackStackEntry = backStackEntry,
    )
  }
}

private fun Flow<List<NavBackStackEntry>>.zipWithPreviousCount(): Flow<Pair<Int, List<NavBackStackEntry>>> = flow {
  var previous = 0
  collect { value ->
    emit(Pair(previous, value))
    previous = value.count()
  }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun ModalBottomSheet(
  navigator: BottomSheetNavigator,
  saveableStateHolder: SaveableStateHolder,
  targetBackStackEntry: NavBackStackEntry?,
) {
  val sheetState = rememberModalBottomSheetState()
  val backStackEntry by produceState<NavBackStackEntry?>(
    initialValue = null,
    key1 = targetBackStackEntry,
  ) {
    try {
      sheetState.hide()
    } catch (_: CancellationException) {
      // We catch but ignore possible cancellation exceptions as we don't want
      // them to bubble up and cancel the whole produceState coroutine
    } finally {
      value = targetBackStackEntry
    }
  }

  ModalBototmSheet(
    sheetState = sheetState,
    navigator = navigator,
    saveableStateHolder = saveableStateHolder,
    backStackEntry = backStackEntry ?: return,
  )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun ModalBototmSheet(
  sheetState: SheetState,
  navigator: BottomSheetNavigator,
  saveableStateHolder: SaveableStateHolder,
  backStackEntry: NavBackStackEntry,
) {
  LaunchedEffect(backStackEntry) {
    sheetState.show()
  }

  backStackEntry.LocalOwnersProvider(saveableStateHolder) {
    val destination = backStackEntry.destination as BottomSheetNavigator.Destination

    ModalBottomSheet(
      sheetState = sheetState,
      onDismissRequest = { navigator.dismiss(backStackEntry) },
      shape = destination.sheetProperties.shape,
      containerColor = destination.sheetProperties.containerColor,
      contentColor = destination.sheetProperties.contentColor,
      tonalElevation = destination.sheetProperties.tonalElevation,
      scrimColor = destination.sheetProperties.scrimColor,
      dragHandle = destination.dragHandle ?: { DragHandle() },
    ) {
      LaunchedEffect(backStackEntry) {
        navigator.onTransitionComplete(backStackEntry)
      }
      DisposableEffect(backStackEntry) {
        onDispose {
          navigator.onTransitionComplete(backStackEntry)
        }
      }
      destination.content(this, backStackEntry)
    }
  }
}
```

Reply:
Personally, I use the current `sheetState.hide()` function to handle the dismiss animation by passing the sheetState as a parameter.

If this was handled automatically as you've done above, that would simplify a lot of "plaster" code many people probably have ;)

Thanks for your work on this, it's hopefully going to be a great help!